File: sol/2594/2594.cpp
struct Solution {
    using ll = long long;
    long long repairCars(vector<int>& ranks, int cars) {
        ranges::sort(ranks);
        ll s = 1, e = static_cast<ll>(ranks[0]) * cars * cars;
        while (s < e){
            ll m = s + (e - s) / 2;
            int remain = cars;
            for (int r : ranks){
                remain -= static_cast<int>(sqrt(m / r));
                if (remain <= 0){
                    break;
                }
            }
            if (remain <= 0){
                e = m;
            } else {
                s = m + 1;
            }
        }
        return s;
    }
};

File: sol/2594/2594.py
class Solution:
    def repairCars(self, ranks: List[int], cars: int) -> int:
        ranks.sort()
        s = ranks[0] * (cars // len(ranks)) ** 2
        e = ranks[0] * cars ** 2
        while s < e:
            m = (s + e) // 2
            remain = cars
            for r in ranks:
                remain -= int((m / r) ** 0.5)
                if remain <= 0:
                    break
            if remain <= 0:
                e = m
            else:
                s = m + 1
        return s
